clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2412,6 +2412,19 @@ static bool isESIMDKernelType(const CXXRecordDecl *KernelObjType) {
   return (OpParens != nullptr) && OpParens->hasAttr<SYCLSimdAttr>();
 }
 
+// Fetch the associated call operator of the kernel object
+// (of either the lambda or the function object).
+static bool IsCallOperatorDefined(const CXXRecordDecl *KernelObjType) {
+  const CXXMethodDecl *CallOperator = nullptr;
+  if (!KernelObjType)
+    return false;
+  if (KernelObjType->isLambda())
+    CallOperator = KernelObjType->getLambdaCallOperator();
+  else
+    CallOperator = getOperatorParens(KernelObjType);
+  return (CallOperator != nullptr);
+}
+
```

Comment:
Looks good to me.

Around line 2512, there is identical logic used for getting the call operator.   Can you refactor this function (perhaps by having it return the CallOperator) so it can be called there as well?